crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -119,6 +121,63 @@ impl Solution {
         &self.score
     }
 
+    /// JIT score calculation as per CIP38
+    pub fn scoring(&self, prices: &auction::Prices) -> Result<eth::TokenAmount, ScoringError> {
+        let mut trades = Vec::with_capacity(self.trades.len());
+        for trade in self.user_trades() {
+            // Solver generated fulfillment does not include the fee in the executed amount
+            // for sell orders.
+            let executed = match trade.order().side {
+                order::Side::Sell => (trade.executed().0 + trade.fee().0).into(),
+                order::Side::Buy => trade.executed(),
+            };
+            let uniform_prices = ClearingPrices {
+                sell: *self
+                    .prices
+                    .get(&trade.order().sell.token.wrap(self.weth))
+                    .ok_or(SolutionError::InvalidClearingPrices)?,
+                buy: *self
+                    .prices
+                    .get(&trade.order().buy.token.wrap(self.weth))
+                    .ok_or(SolutionError::InvalidClearingPrices)?,
+            };
+            let custom_prices = scoring::CustomClearingPrices {
+                sell: match trade.order().side {
+                    order::Side::Sell => trade
+                        .executed()
+                        .0
+                        .checked_mul(uniform_prices.sell)
+                        .ok_or(Math::Overflow)?
+                        .checked_div(uniform_prices.buy)
+                        .ok_or(Math::DivisionByZero)?,
+                    order::Side::Buy => trade.executed().0,
+                },
+                buy: match trade.order().side {
+                    order::Side::Sell => trade.executed().0 + trade.fee().0,
+                    order::Side::Buy => {
+                        (trade.executed().0)
+                            .checked_mul(uniform_prices.buy)
+                            .ok_or(Math::Overflow)?
+                            .checked_div(uniform_prices.sell)
+                            .ok_or(Math::DivisionByZero)?
+                            + trade.fee().0
```

Comment:
> Then one could probably even think about merging the two Trade domain models

We slowly converged to calculating score on fulfillment trade. :) However, this was not my intention at first. Initially, I wanted to do scoring strictly on different type, a type that would be final after all driver operations done on the solution received from solver. But since we are moving away from driver making any changes to the solver solutions, I guess this approach makes sense until we are careful about adding any features that could alter the score.